### PR TITLE
Revert tile drawing optimizations

### LINKED ISSUE #3539: View Bugs (1.9.0.0.10535)
Intel Core i5, 7th generation.
8 GB RAM.
Windows 10.
TripleA 1.9.0.0.10535 installed on SSD (TripleA.vmoptions: -Xmx2G -Xms2G).
world_at_war-master.zip on HDD.

File Video (thanks to @RoiEXLab for uploading it on the RoiEX channel):
https://youtu.be/PpgGGC3d5ks

00:25. I try to change the zoom from 10 to 100, but it doesn't work (eventually, I zoom using the mouse wheel, instead).
01:06. Since the borders are drawn medium when loading under 100 zoom (due to this issue: https://github.com/triplea-game/triplea/issues/3525), I try to change to low, partially failing, and getting an error.
01:59. I start the game at 100 zoom, but the map is not readily drawn in all cases outside the starting view.
02:35. I set the zoom at 10, to have all the map on view, but it will take until 10:51 to complete drawing (so more than 8 minutes to fully draw the map).
11:38. I move some units to attack some territory, but they are not redrawn in the target territory (I believe this happened only after the changes eliminating the flickering issues when moving units).

### LINKED ISSUE #3707: Show Map Blends command no longer works
### Engine version
1.9.0.0.11026

### My Operating System
Linux localhost.localdomain 4.17.3-200.fc28.x86_64 <span>#</span>1 SMP Tue Jun 26 14:17:07 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

### Can you describe how to trigger the error? (eg: what sequence of actions will recreate it?)

1. Start a local game.
1. Select **View > Show Map Blends**.

### Do you have the exact error text? Please copy/paste if so

Map blends are not drawn when the menu item is checked.

### Instead of this error, what should have happened?

Map blends should be drawn.

### Any additional information that may help

If you leave the Show Map Blends menu item checked, leave the game, and return, the map blends will be drawn.  That suggests the map blend code itself is working correctly but whatever triggers it is in response to the menu item is not.

I bisected the first occurrence of this failure to 1.9.0.0.10509 (#3482).

### PR DESCRIPTION
**Pull request by ssoloff**
## Overview

~Potentially~ fixes #3539; fixes #3707.

## Functional Changes

The goal here was to revert all tile drawing optimizations going back until 1.9.0.0.10506, which was the last known good build.  The following PRs were reverted:

* #3482
* #3519
* #3521
* #3528
* #3532
* #3538
* #3567
* #3583
* #3625
* #3655

## Manual Testing Performed

* Smoke tested a few maps to ensure tiles were rendered as expected.
* Verified map blends are now rendered/not rendered when using the Show Map Blends command (#3707).
* :warning: As mentioned elsewhere, I was unable to reproduce the issue described in #3539.  Therefore, I am unable to determine if that issue truly is resolved.  I plan on building a complete release in my fork and will ask users to test that build shortly; dev testing of this branch is also welcome.  **Please do not merge until the TESTING label is removed.**

## Additional Review notes

I attempted to preserve the changes from all other PRs that touched the affected files.  In addition to reviewing the changes of this PR, reviewers should review the following diff, which compares the files changed in this PR to 1.9.0.0.10506.

<details>
<summary><tt>git diff --name-only master HEAD | xargs git diff 1.9.0.0.10506 --</tt></summary>

```diff
diff --git a/game-core/gradle.properties b/game-core/gradle.properties
index 7bafbd893..308a34e43 100644
--- a/game-core/gradle.properties
+++ b/game-core/gradle.properties
@@ -1,2 +1,2 @@
-checkstyleMainMaxWarnings=1584
-checkstyleTestMaxWarnings=4
+checkstyleMainMaxWarnings=1280
+checkstyleTestMaxWarnings=2
diff --git a/game-core/src/main/java/games/strategy/triplea/image/TileImageFactory.java b/game-core/src/main/java/games/strategy/triplea/image/TileImageFactory.java
index 676031c69..7073d5ab7 100644
--- a/game-core/src/main/java/games/strategy/triplea/image/TileImageFactory.java
+++ b/game-core/src/main/java/games/strategy/triplea/image/TileImageFactory.java
@@ -1,7 +1,5 @@
 package games.strategy.triplea.image;
 
-import java.awt.AlphaComposite;
-import java.awt.Composite;
 import java.awt.Graphics;
 import java.awt.Graphics2D;
 import java.awt.GraphicsConfiguration;
@@ -39,7 +37,6 @@ public final class TileImageFactory {
   private static String showMapBlendMode;
   private static final String SHOW_MAP_BLEND_ALPHA = "BlendAlpha";
   private static float showMapBlendAlpha;
-  private final Composite composite = AlphaComposite.Src;
   private static final GraphicsConfiguration configuration =
       GraphicsEnvironment.getLocalGraphicsEnvironment().getDefaultScreenDevice().getDefaultConfiguration();
   private static final Logger logger = Logger.getLogger(TileImageFactory.class.getName());
@@ -192,7 +189,7 @@ public final class TileImageFactory {
   }
 
   /**
-   * @return compatibleImage This method produces a blank white tile for use in blending.
+   * This method produces a blank white tile for use in blending.
    */
   private static BufferedImage makeMissingBaseTile(final BufferedImage input) {
     final BufferedImage compatibleImage =
@@ -316,10 +313,6 @@ public final class TileImageFactory {
     return image;
   }
 
-  public Composite getComposite() {
-    return this.composite;
-  }
-
   private static BufferedImage loadCompatibleImage(final URL resource) throws IOException {
     final BufferedImage image = ImageIO.read(resource);
     return toCompatibleImage(image);
diff --git a/game-core/src/main/java/games/strategy/triplea/ui/HeadedUiContext.java b/game-core/src/main/java/games/strategy/triplea/ui/HeadedUiContext.java
index f38a43c18..e69740cbf 100644
--- a/game-core/src/main/java/games/strategy/triplea/ui/HeadedUiContext.java
+++ b/game-core/src/main/java/games/strategy/triplea/ui/HeadedUiContext.java
@@ -142,7 +142,7 @@ public class HeadedUiContext extends AbstractUiContext {
       final UnitDamage damaged, final UnitEnable disabled) {
     final Optional<ImageIcon> image = getUnitImageFactory().getIcon(type, player, damaged == UnitDamage.DAMAGED,
         disabled == UnitEnable.DISABLED);
-    JLabel label = image.map(JLabel::new).orElseGet(JLabel::new);
+    final JLabel label = image.map(JLabel::new).orElseGet(JLabel::new);
     MapUnitTooltipManager.setUnitTooltip(label, type, player, 1);
     return label;
   }
diff --git a/game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java b/game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java
index fdd65134e..e79803002 100644
--- a/game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java
+++ b/game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java
@@ -9,7 +9,9 @@ import java.awt.Dimension;
 import java.awt.Graphics;
 import java.awt.Graphics2D;
 import java.awt.Image;
+import java.awt.MouseInfo;
 import java.awt.Point;
+import java.awt.PointerInfo;
 import java.awt.Rectangle;
 import java.awt.RenderingHints;
 import java.awt.event.MouseAdapter;
@@ -187,28 +189,20 @@ public class MapPanel extends ImageScrollerLargeView {
     addMouseMotionListener(new MouseMotionAdapter() {
       @Override
       public void mouseMoved(final MouseEvent e) {
-        final MouseDetails md = convert(e);
-        final double scaledMouseX = e.getX() / scale;
-        final double scaledMouseY = e.getY() / scale;
-        final double x = normalizeX(scaledMouseX + getXOffset());
-        final double y = normalizeY(scaledMouseY + getYOffset());
-        final Territory terr = getTerritory(x, y);
-        if (!Objects.equals(terr, currentTerritory)) {
-          currentTerritory = terr;
-          notifyMouseEntered(terr);
-        }
-        notifyMouseMoved(terr, md);
-        final Tuple<Territory, List<Unit>> tuple = tileManager.getUnitsAtPoint(x, y, gameData);
-        if (unitsChanged(tuple)) {
-          currentUnits = tuple;
-          if (tuple == null) {
-            notifyMouseEnterUnit(Collections.emptyList(), getTerritory(x, y), md);
-          } else {
-            notifyMouseEnterUnit(tuple.getSecond(), tuple.getFirst(), md);
-          }
-        }
+        updateMouseHoverState(convert(e), e.getX(), e.getY());
       }
     });
+    // When map is scrolled, update information about what we're hovering over.
+    model.addObserver((object, arg) -> {
+      SwingUtilities.invokeLater(() -> {
+        final PointerInfo pointer = MouseInfo.getPointerInfo();
+        if (pointer != null) {
+          final Point loc = pointer.getLocation();
+          SwingUtilities.convertPointFromScreen(loc, MapPanel.this);
+          updateMouseHoverState(null, loc.x, loc.y);
+        }
+      });
+    });
     addScrollListener((x2, y2) -> SwingUtilities.invokeLater(this::repaint));
     executor.submit(() -> recreateTiles(data, uiContext));
     uiContext.addActive(() -> {
@@ -219,6 +213,30 @@ public class MapPanel extends ImageScrollerLargeView {
     });
   }
 
+  private void updateMouseHoverState(final MouseDetails md, final int mouseX, final int mouseY) {
+    final double scaledMouseX = mouseX / scale;
+    final double scaledMouseY = mouseY / scale;
+    final double x = normalizeX(scaledMouseX + getXOffset());
+    final double y = normalizeY(scaledMouseY + getYOffset());
+    final Territory terr = getTerritory(x, y);
+    if (!Objects.equals(terr, currentTerritory)) {
+      currentTerritory = terr;
+      notifyMouseEntered(terr);
+    }
+    if (md != null) {
+      notifyMouseMoved(terr, md);
+    }
+    final Tuple<Territory, List<Unit>> tuple = tileManager.getUnitsAtPoint(x, y, gameData);
+    if (unitsChanged(tuple)) {
+      currentUnits = tuple;
+      if (tuple == null) {
+        notifyMouseEnterUnit(Collections.emptyList(), getTerritory(x, y), md);
+      } else {
+        notifyMouseEnterUnit(tuple.getSecond(), tuple.getFirst(), md);
+      }
+    }
+  }
+
   private void recreateTiles(final GameData data, final UiContext uiContext) {
     this.tileManager.createTiles(new Rectangle(this.uiContext.getMapData().getMapDimensions()));
     this.tileManager.resetTiles(data, uiContext.getMapData());
@@ -358,28 +376,28 @@ public class MapPanel extends ImageScrollerLargeView {
     return gameData.getMap().getTerritory(name);
   }
 
-  private double normalizeX(double x) {
+  private double normalizeX(final double x) {
     if (!uiContext.getMapData().scrollWrapX()) {
       return x;
     }
     final int imageWidth = (int) getImageDimensions().getWidth();
     if (x < 0) {
-      x += imageWidth;
+      return x + imageWidth;
     } else if (x > imageWidth) {
-      x -= imageWidth;
+      return x - imageWidth;
     }
     return x;
   }
 
-  private double normalizeY(double y) {
+  private double normalizeY(final double y) {
     if (!uiContext.getMapData().scrollWrapY()) {
       return y;
     }
     final int imageHeight = (int) getImageDimensions().getHeight();
     if (y < 0) {
-      y += imageHeight;
+      return y + imageHeight;
     } else if (y > imageHeight) {
-      y -= imageHeight;
+      return y - imageHeight;
     }
     return y;
   }
diff --git a/game-core/src/main/java/games/strategy/triplea/ui/logic/RouteCalculator.java b/game-core/src/main/java/games/strategy/triplea/ui/logic/RouteCalculator.java
index e23228543..e4abab207 100644
--- a/game-core/src/main/java/games/strategy/triplea/ui/logic/RouteCalculator.java
+++ b/game-core/src/main/java/games/strategy/triplea/ui/logic/RouteCalculator.java
@@ -15,8 +15,6 @@ import java.util.stream.Collectors;
 
 import javax.annotation.Nullable;
 
-import com.google.common.annotations.VisibleForTesting;
-
 import lombok.AccessLevel;
 import lombok.AllArgsConstructor;
 import lombok.Builder;
diff --git a/game-core/src/main/java/games/strategy/triplea/ui/menubar/ViewMenu.java b/game-core/src/main/java/games/strategy/triplea/ui/menubar/ViewMenu.java
index 13bf368a0..95c8b8be1 100644
--- a/game-core/src/main/java/games/strategy/triplea/ui/menubar/ViewMenu.java
+++ b/game-core/src/main/java/games/strategy/triplea/ui/menubar/ViewMenu.java
@@ -42,7 +42,6 @@ import games.strategy.engine.data.properties.ColorProperty;
 import games.strategy.engine.data.properties.IEditableProperty;
 import games.strategy.engine.data.properties.NumberProperty;
 import games.strategy.engine.data.properties.PropertiesUi;
-import games.strategy.engine.framework.GameRunner;
 import games.strategy.triplea.image.MapImage;
 import games.strategy.triplea.image.TileImageFactory;
 import games.strategy.triplea.ui.AbstractUiContext;
diff --git a/game-core/src/main/java/games/strategy/triplea/ui/screen/TerritoryOverLayDrawable.java b/game-core/src/main/java/games/strategy/triplea/ui/screen/TerritoryOverLayDrawable.java
index 7031336ad..8f129b725 100644
--- a/game-core/src/main/java/games/strategy/triplea/ui/screen/TerritoryOverLayDrawable.java
+++ b/game-core/src/main/java/games/strategy/triplea/ui/screen/TerritoryOverLayDrawable.java
@@ -11,6 +11,7 @@ import games.strategy.engine.data.GameData;
 import games.strategy.engine.data.Territory;
 import games.strategy.triplea.ui.mapdata.MapData;
 import games.strategy.triplea.ui.screen.drawable.IDrawable;
+import games.strategy.ui.Util;
 
 class TerritoryOverLayDrawable implements IDrawable {
   enum Operation {
@@ -39,18 +40,16 @@ class TerritoryOverLayDrawable implements IDrawable {
     final Territory territory = data.getMap().getTerritory(territoryName);
     final List<Polygon> polys = mapData.getPolygons(territory);
     graphics.setColor(color);
-    for (Polygon polygon : polys) {
-      // if we dont have to draw, dont
+    for (final Polygon polygon : polys) {
       if (!polygon.intersects(bounds) && !polygon.contains(bounds)) {
         continue;
       }
-      // use a copy since we will move the polygon
-      polygon = new Polygon(polygon.xpoints, polygon.ypoints, polygon.npoints);
-      polygon.translate(-bounds.x, -bounds.y);
+
+      final Polygon translatedPolygon = Util.translatePolygon(polygon, -bounds.x, -bounds.y);
       if (operation == Operation.FILL) {
-        graphics.fillPolygon(polygon);
+        graphics.fillPolygon(translatedPolygon);
       } else {
-        graphics.drawPolygon(polygon);
+        graphics.drawPolygon(translatedPolygon);
       }
     }
   }
diff --git a/game-core/src/main/java/games/strategy/triplea/ui/screen/TileManager.java b/game-core/src/main/java/games/strategy/triplea/ui/screen/TileManager.java
index d088f9438..19d6466ff 100644
--- a/game-core/src/main/java/games/strategy/triplea/ui/screen/TileManager.java
+++ b/game-core/src/main/java/games/strategy/triplea/ui/screen/TileManager.java
@@ -182,14 +182,14 @@ public class TileManager {
     try {
       acquireLock();
       try {
-        for (Tile tile : tiles) {
+        for (final Tile tile : tiles) {
           tile.clear();
           final int x = tile.getBounds().x / TILE_SIZE;
           final int y = tile.getBounds().y / TILE_SIZE;
           tile.addDrawable(new BaseMapDrawable(x, y, uiContext));
           tile.addDrawable(new ReliefMapDrawable(x, y, uiContext));
         }
-        for (Territory territory : data.getMap().getTerritories()) {
+        for (final Territory territory : data.getMap().getTerritories()) {
           clearTerritory(territory);
           drawTerritory(territory, data, mapData);
         }
@@ -221,7 +221,7 @@ public class TileManager {
         if (territories == null) {
           return;
         }
-        for (Territory territory : territories) {
+        for (final Territory territory : territories) {
           updateTerritory(territory, data, mapData);
         }
       } finally {
@@ -256,7 +256,7 @@ public class TileManager {
     if (drawables == null || drawables.isEmpty()) {
       return;
     }
-    for (Tile tile : territoryTiles.get(territory.getName())) {
+    for (final Tile tile : territoryTiles.get(territory.getName())) {
       tile.removeDrawables(drawables);
     }
     allUnitDrawables.removeAll(drawables);
@@ -312,7 +312,7 @@ public class TileManager {
       drawing.add(new VcDrawable(territory));
     }
     // add to the relevant tiles
-    for (Tile tile : getTiles(mapData.getBoundingRect(territory.getName()))) {
+    for (final Tile tile : getTiles(mapData.getBoundingRect(territory.getName()))) {
       drawnOn.add(tile);
       tile.addDrawables(drawing);
     }
@@ -357,7 +357,7 @@ public class TileManager {
           category.getDisabled(), overflow, territory.getName(), uiContext);
       drawing.add(drawable);
       allUnitDrawables.add(drawable);
-      for (Tile tile : getTiles(
+      for (final Tile tile : getTiles(
           new Rectangle(lastPlace.x, lastPlace.y, uiContext.getUnitImageFactory().getUnitImageWidth(),
               uiContext.getUnitImageFactory().getUnitImageHeight()))) {
         tile.addDrawable(drawable);
@@ -518,10 +518,8 @@ public class TileManager {
     }
     graphics.setStroke(new BasicStroke(10));
     graphics.setColor(Color.RED);
-    for (Polygon poly : mapData.getPolygons(selected)) {
-      poly = new Polygon(poly.xpoints, poly.ypoints, poly.npoints);
-      poly.translate(-bounds.x, -bounds.y);
-      graphics.drawPolygon(poly);
+    for (final Polygon polygon : mapData.getPolygons(selected)) {
+      graphics.drawPolygon(Util.translatePolygon(polygon, -bounds.x, -bounds.y));
     }
   }
 
diff --git a/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/LandTerritoryDrawable.java b/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/LandTerritoryDrawable.java
index d902227de..499873b57 100644
--- a/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/LandTerritoryDrawable.java
+++ b/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/LandTerritoryDrawable.java
@@ -20,14 +20,26 @@ public class LandTerritoryDrawable extends TerritoryDrawable implements IDrawabl
   @Override
   public void draw(final Rectangle bounds, final GameData data, final Graphics2D graphics, final MapData mapData,
       final AffineTransform unscaled, final AffineTransform scaled) {
+    draw(bounds, data, graphics, mapData, 1.0f);
+  }
+
+  /**
+   * Determine territory color and set saturation to then draw the territory.
+   */
+  public void draw(final Rectangle bounds, final GameData data, final Graphics2D graphics, final MapData mapData,
+      final float saturation) {
     final Territory territory = data.getMap().getTerritory(territoryName);
-    final Color territoryColor;
+    Color territoryColor;
     final TerritoryAttachment ta = TerritoryAttachment.get(territory);
     if (ta != null && ta.getIsImpassable()) {
       territoryColor = mapData.impassableColor();
     } else {
       territoryColor = mapData.getPlayerColor(territory.getOwner().getName());
     }
+    final float[] values =
+        Color.RGBtoHSB(territoryColor.getRed(), territoryColor.getGreen(), territoryColor.getBlue(), null);
+    values[1] = values[1] * saturation;
+    territoryColor = Color.getHSBColor(values[0], values[1], values[2]);
     draw(bounds, graphics, mapData, territory, territoryColor);
   }
 
diff --git a/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/OptionalExtraTerritoryBordersDrawable.java b/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/OptionalExtraTerritoryBordersDrawable.java
index b60ac5fb2..b5c145354 100644
--- a/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/OptionalExtraTerritoryBordersDrawable.java
+++ b/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/OptionalExtraTerritoryBordersDrawable.java
@@ -10,6 +10,7 @@ import java.util.List;
 import games.strategy.engine.data.GameData;
 import games.strategy.engine.data.Territory;
 import games.strategy.triplea.ui.mapdata.MapData;
+import games.strategy.ui.Util;
 
 public class OptionalExtraTerritoryBordersDrawable implements IDrawable {
   private final String territoryName;
@@ -25,16 +26,13 @@ public class OptionalExtraTerritoryBordersDrawable implements IDrawable {
       final AffineTransform unscaled, final AffineTransform scaled) {
     final Territory territory = data.getMap().getTerritory(territoryName);
     final List<Polygon> polys = mapData.getPolygons(territory);
-    for (Polygon polygon : polys) {
-      // if we dont have to draw, dont
+    graphics.setColor(Color.BLACK);
+    for (final Polygon polygon : polys) {
       if (!polygon.intersects(bounds) && !polygon.contains(bounds)) {
         continue;
       }
-      // use a copy since we will move the polygon
-      polygon = new Polygon(polygon.xpoints, polygon.ypoints, polygon.npoints);
-      polygon.translate(-bounds.x, -bounds.y);
-      graphics.setColor(Color.BLACK);
-      graphics.drawPolygon(polygon);
+
+      graphics.drawPolygon(Util.translatePolygon(polygon, -bounds.x, -bounds.y));
     }
   }
 
diff --git a/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/SeaZoneOutlineDrawable.java b/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/SeaZoneOutlineDrawable.java
index bf05eff04..54e039440 100644
--- a/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/SeaZoneOutlineDrawable.java
+++ b/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/SeaZoneOutlineDrawable.java
@@ -10,6 +10,7 @@ import java.util.List;
 import games.strategy.engine.data.GameData;
 import games.strategy.engine.data.Territory;
 import games.strategy.triplea.ui.mapdata.MapData;
+import games.strategy.ui.Util;
 
 public class SeaZoneOutlineDrawable implements IDrawable {
   private final String territoryName;
@@ -23,16 +24,13 @@ public class SeaZoneOutlineDrawable implements IDrawable {
       final AffineTransform unscaled, final AffineTransform scaled) {
     final Territory territory = data.getMap().getTerritory(territoryName);
     final List<Polygon> polys = mapData.getPolygons(territory);
-    for (Polygon polygon : polys) {
-      // if we dont have to draw, dont
+    graphics.setColor(Color.BLACK);
+    for (final Polygon polygon : polys) {
       if (!polygon.intersects(bounds) && !polygon.contains(bounds)) {
         continue;
       }
-      // use a copy since we will move the polygon
-      polygon = new Polygon(polygon.xpoints, polygon.ypoints, polygon.npoints);
-      polygon.translate(-bounds.x, -bounds.y);
-      graphics.setColor(Color.BLACK);
-      graphics.drawPolygon(polygon);
+
+      graphics.drawPolygon(Util.translatePolygon(polygon, -bounds.x, -bounds.y));
     }
   }
 
diff --git a/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java b/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java
index b7e297ce4..e33298721 100644
--- a/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java
+++ b/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java
@@ -22,6 +22,7 @@ import games.strategy.triplea.ui.mapdata.MapData;
 public class TerritoryNameDrawable implements IDrawable {
   private final String territoryName;
   private final UiContext uiContext;
+  private Rectangle territoryBounds;
 
   public TerritoryNameDrawable(final String territoryName, final UiContext uiContext) {
     this.territoryName = territoryName;
@@ -82,7 +83,10 @@ public class TerritoryNameDrawable implements IDrawable {
       x = namePlace.get().x;
       y = namePlace.get().y;
     } else {
-      final Rectangle territoryBounds = getBestTerritoryNameRect(mapData, territory, fm);
+      if (territoryBounds == null) {
+        // Cache the bounds since re-computing it is expensive.
+        territoryBounds = getBestTerritoryNameRect(mapData, territory, fm);
+      }
       x = territoryBounds.x + (int) territoryBounds.getWidth() / 2 - fm.stringWidth(territory.getName()) / 2;
       y = territoryBounds.y + (int) territoryBounds.getHeight() / 2 + fm.getAscent() / 2;
     }
diff --git a/game-core/src/main/java/games/strategy/ui/ImageScrollerLargeView.java b/game-core/src/main/java/games/strategy/ui/ImageScrollerLargeView.java
index ea105553a..0c8171c44 100644
--- a/game-core/src/main/java/games/strategy/ui/ImageScrollerLargeView.java
+++ b/game-core/src/main/java/games/strategy/ui/ImageScrollerLargeView.java
@@ -300,6 +300,8 @@ public class ImageScrollerLargeView extends JComponent {
   }
 
   /**
+   * Sets the view scale.
+   *
    * @param value The new scale value. Constrained to the bounds of no less than 0.15 and no greater than 1.
    *        If out of bounds the nearest boundary value is used.
    */
diff --git a/game-core/src/test/java/games/strategy/triplea/ui/logic/RouteCalculatorTest.java b/game-core/src/test/java/games/strategy/triplea/ui/logic/RouteCalculatorTest.java
index 855befcd9..3b08ea2f5 100644
--- a/game-core/src/test/java/games/strategy/triplea/ui/logic/RouteCalculatorTest.java
+++ b/game-core/src/test/java/games/strategy/triplea/ui/logic/RouteCalculatorTest.java
@@ -17,8 +17,7 @@ import java.util.List;
 
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.ExtendWith;
-
-import com.example.mockito.MockitoExtension;
+import org.mockito.junit.jupiter.MockitoExtension;
 
 @ExtendWith(MockitoExtension.class)
 public class RouteCalculatorTest {
```
</details>

### CODE REVIEW

File: game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java [hunk 1]
```diff
@@ -205,7 +204,7 @@ public void mouseMoved(final MouseEvent e) {
       });
     });
     addScrollListener((x2, y2) -> SwingUtilities.invokeLater(this::repaint));
-    executor.execute(() -> recreateTiles(data, uiContext));
+    executor.submit(() -> recreateTiles(data, uiContext));
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java]: This should stay execute, otherwise any errors might get swallowed

File: game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java [hunk 2]
```diff
@@ -292,7 +288,7 @@ public void centerOn(final Territory territory) {
       return;
     }
     final Point p = uiContext.getMapData().getCenter(territory);
-    // when centering don't want the map to wrap around,
+    // when centering dont want the map to wrap around,
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java]: Typo

File: game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java [hunk 3]
```diff
@@ -447,27 +443,27 @@ void setGameData(final GameData data) {
     gameData.addTerritoryListener(territoryListener);
     gameData.addDataChangeListener(dataChangeListener);
     clearPendingDrawOperations();
-    executor.execute(() -> tileManager.resetTiles(gameData, uiContext.getMapData()));
+    executor.submit(() -> tileManager.resetTiles(gameData, uiContext.getMapData()));
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java]: Same here, about swallowed errors

File: game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java [hunk 4]
```diff
@@ -447,27 +443,27 @@ void setGameData(final GameData data) {
...
       updateCountries(Collections.singleton(territory));
-      SwingUtilities.invokeLater(MapPanel.this::repaint);
+      SwingUtilities.invokeLater(() -> repaint());
     }
 
     @Override
     public void attachmentChanged(final Territory territory) {
       updateCountries(Collections.singleton(territory));
-      SwingUtilities.invokeLater(MapPanel.this::repaint);
+      SwingUtilities.invokeLater(() -> repaint());
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java]: Shorter lambda notation can stay as well

File: game-core/src/main/java/games/strategy/triplea/ui/logic/MapScrollUtil.java [hunk 5]
```diff
@@ -1,57 +0,0 @@
...
-import java.awt.geom.AffineTransform;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.List;
-
-/**
- * Helper class to deal with infinitely scrolling maps.
- */
-public class MapScrollUtil {
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/ui/logic/MapScrollUtil.java]: This is the class with the extracted multi-function method I mentioned

File: game-core/src/main/java/games/strategy/triplea/ui/logic/RouteCalculator.java [hunk 6]
```diff
@@ -125,10 +126,45 @@ private static Path2D getNormalizedLines(final double[] xcoords, final double[]
...
+          AffineTransform.getTranslateInstance(-mapWidth, 0),
+          AffineTransform.getTranslateInstance(+mapWidth, 0)));
+    }
+    if (isInfiniteY) {
+      result.addAll(Arrays.asList(
+          AffineTransform.getTranslateInstance(0, -mapHeight),
+          AffineTransform.getTranslateInstance(0, +mapHeight)));
+    }
+    return Collections.unmodifiableList(result);
+  }
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/ui/logic/RouteCalculator.java]: Here the method to extract

File: game-core/src/main/java/games/strategy/triplea/ui/screen/Tile.java [hunk 7]
```diff
@@ -1,121 +1,157 @@
...
+    if (scale != 1) {
+      scaled = new AffineTransform();
+      scaled.scale(scale, scale);
+      g.setTransform(scaled);
+    } else {
+      scaled = unscaled;
+    }
     final Stopwatch stopWatch = new Stopwatch(Logger.getLogger(Tile.class.getName()), Level.FINEST,
-        "Drawing Tile at " + bounds);
-    final Queue<IDrawable> clone = new PriorityBlockingQueue<>(contents);
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/ui/screen/Tile.java]: Queue should work as well (99% sure)

File: game-core/src/main/java/games/strategy/triplea/ui/screen/TileManager.java [hunk 8]
```diff
@@ -110,21 +145,12 @@ public TileManager(final UiContext uiContext) {
...
-    try {
-      return new ArrayList<>(tiles);
-    } finally {
-      releaseLock();
-    }
-  }
-
   private void acquireLock() {
-    LockUtil.INSTANCE.acquireLock(lock);
+    Tile.LOCK_UTIL.acquireLock(lock);
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/ui/screen/TileManager.java]: Tile.LOCK_UTIL is a static reference to LockUtil.INSTANCE, so this can be kept as well

File: game-core/src/main/java/games/strategy/triplea/ui/logic/RouteCalculator.java [hunk 9]
```diff
@@ -3,25 +3,28 @@
...
-/**
- * Utility class to calculate a smooth route to draw on the map.
- */
+@AllArgsConstructor(access =  AccessLevel.PUBLIC)
 @Builder
 public class RouteCalculator {
 
-  public final boolean isInfiniteX;
   public final boolean isInfiniteY;
+  public final boolean isInfiniteX;
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/ui/logic/RouteCalculator.java]: The members were reversed

**Comment by ssoloff** [comment]
@RoiEXLab Totally agree with your observations about changes that probably have no effect on #3539.  However, due to the complexity of the revert, I wanted to keep the diff between this branch and 1.9.0.0.10506 as small as possible so I could be sure I got everything.

Once this PR is merged, I will follow-up on some of the issues you've noted.  Anything that I miss will be waiting for you when you get back in three weeks. :smile:

File: game-core/src/main/java/games/strategy/triplea/ui/logic/RouteCalculator.java (hunk 9 above)
Comment (ssoloff) [reply game-core/src/main/java/games/strategy/triplea/ui/logic/RouteCalculator.java]: Because this snapshot is still using `@AllArgsConstructor`, the member declaration order dictates the parameter order in the constructor.  There is one call to the constructor in `ImageScrollerSmallView` that still specifies `isInfiniteY` before `isInfiniteX`, so this field order is technically correct for the 1.9.0.0.10506 snapshot.

Before reordering the fields, the all-args constructor should probably be made private, and the remaining call site in `ImageScrollerSmallView` should be converted to builder usage.

**Comment by ssoloff** [comment]
Just noticed that #3496 was _not_ reverted, as documented above.  Even though it was opened after #3482, it was merged in 1.9.0.0.10466, and wasn't included in the commands I ran to revert changes for this PR.  I think that's actually correct since we've already had confirmation from users testing 1.9.0.0.10506 that that PR didn't cause #3539.

I updated this PR's description to remove #3496 from the list of reverted PRs.

**Comment by ssoloff** [comment]
Also realized that I missed restoring the unrelated changes from #3580 and #3649--only a few lines.  I updated the diff against 1.9.0.0.10506 in the description appropriately.

**Comment by DanVanAtta** [comment]
I agree a clean revert and walking back the changes is hopefully the most efficient/safe way to go. This code base despite looking simple, is not : |

File: game-core/src/main/java/games/strategy/triplea/ui/menubar/ViewMenu.java [hunk 10]
```diff
@@ -127,10 +127,10 @@ private void addShowGameUuid() {
   private void addZoomMenu() {
     final Action mapZoom = SwingAction.of("Map Zoom", e -> {
       final SpinnerNumberModel model = new SpinnerNumberModel();
-      model.setMaximum(200);
-      model.setMinimum(10);
+      model.setMaximum(100);
+      model.setMinimum(15);
       model.setStepSize(1);
-      model.setValue((int)Math.round(frame.getMapPanel().getScale() * 100));
+      model.setValue((int) (frame.getMapPanel().getScale() * 100));
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/ui/menubar/ViewMenu.java]: Oh and this Bugfix should stay as well.

**Comment by ssoloff** [comment]
Two users have confirmed that #3539 is fixed with these changes.  I think this PR is ready to merge if there are no other comments.

**Comment by ron-murhammer** [comment]
@ssoloff Yeah, @Heppisorus just tested it as well and works much better for him than the latest pre-release. Merging so we can get the pre-release back to a functional state though looks like we have another lobby issue with the column changes that were recently made.
